// This file generated for packaging, do not edit.
// To edit, see original source:
// /home/runner/work/farm-ng-amiga/farm-ng-amiga/protos/farm_ng/canbus/canbus.proto.
//
// Copyright (c) farm-ng, inc.
//
// Licensed under the Amiga Development Kit License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://github.com/farm-ng/amiga-dev-kit/blob/main/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.



syntax = "proto3";

package farm_ng.canbus.proto;

service CanbusService {
  rpc streamCanbusMessages(StreamCanbusRequest)
      returns (stream StreamCanbusReply) {}
  rpc sendCanbusMessage(stream SendCanbusMessageRequest)
      returns (SendCanbusMessageReply) {}
  rpc getServiceState(GetServiceStateRequest) returns (GetServiceStateResult) {}
  rpc startService(StartServiceRequest) returns (StartServiceResult) {}
  rpc stopService(StopServiceRequest) returns (StopServiceResult) {}
  rpc pauseService(PauseServiceRequest) returns (PauseServiceResult) {}
}

enum ReplyStatus {
  OK = 0;
  FAILED = 1;
}

enum CanbusServiceState {
  UNKNOWN = 0;
  STOPPED = 1;
  RUNNING = 2;
  IDLE = 3;
  UNAVAILABLE = 4;
  ERROR = 5;
}

message StreamCanbusRequest {
  string message = 1;
}

message StreamCanbusReply {
  ReplyStatus status = 1;
  RawCanbusMessages messages = 2;
}

message SendCanbusMessageRequest {
  RawCanbusMessage message = 1;
}

message SendCanbusMessageReply {
  ReplyStatus status = 1;
}

message RawCanbusMessage {
  // The time the message is "sent" on the canbus
  // in the steady_clock/time.montonic() time domain
  double stamp = 1;
  uint32 id = 2;
  bool error = 3;
  bool remote_transmission = 4;
  bytes data = 5;
}

message RawCanbusMessages {
  repeated RawCanbusMessage messages = 1;
}

message StartServiceRequest {
  string message = 1;
}

message StopServiceRequest {
  string message = 1;
}

message PauseServiceRequest {
  string message = 1;
}

message StartServiceResult {
  string message = 1;
  ReplyStatus status = 2;
}

message StopServiceResult {
  string message = 1;
  ReplyStatus status = 2;
}

message PauseServiceResult {
  string message = 1;
  ReplyStatus status = 2;
}

message GetServiceStateRequest {
  string message = 1;
}

message GetServiceStateResult {
  string state_name = 1;
  CanbusServiceState state = 2;
  ReplyStatus status = 3;
}
